pkg/sources/forward/data_forward.go

```diff
@@ -0,0 +1,621 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package forward
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/forward"
+	"github.com/numaproj/numaflow/pkg/forward/applier"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/metrics"
+	"github.com/numaproj/numaflow/pkg/shared/idlehandler"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"github.com/numaproj/numaflow/pkg/watermark/fetch"
+	"github.com/numaproj/numaflow/pkg/watermark/publish"
+	"github.com/numaproj/numaflow/pkg/watermark/wmb"
+)
+
+// DataForward reads data from source and forwards to inter-step buffers
+type DataForward struct {
+	// I have my reasons for overriding the default principle https://github.com/golang/go/issues/22602
+	ctx context.Context
+	// cancelFn cancels our new context, our cancellation is little more complex and needs to be well orchestrated, hence
+	// we need something more than a cancel().
+	cancelFn context.CancelFunc
+	// reader reads data from source.
+	reader isb.BufferReader
+	// toBuffers store the toVertex name to its owned buffers mapping.
+	toBuffers          map[string][]isb.BufferWriter
+	toWhichStepDecider forward.ToWhichStepDecider
+	transformer        applier.MapApplier
+	wmFetcher          fetch.Fetcher
+	// toVertexWMPublishers stores the toVertex to publisher mapping.
+	toVertexWMPublishers map[string]publish.Publisher
+	// srcWMPublisher is used to publish source watermark.
+	srcWMPublisher isb.SourceWatermarkPublisher
+	opts           options
+	vertexName     string
+	pipelineName   string
+	// idleManager manages the idle watermark status.
+	idleManager *wmb.IdleManager
+	// wmbChecker checks if the idle watermark is valid.
```

Comment:
I think idleManager is still used by source forwarder because after transformation we can have conditional forwarding to toVertices. Correct me if I am wrong. cc: @jy4096